src/shared/edit-util.c

```diff
@@ -18,18 +18,100 @@
 #include "strv.h"
 #include "tmpfile-util.h"
 
-void edit_file_free_all(EditFile **ef) {
-        assert(ef);
+static void edit_file_context_unlink_all(EditFileContext *context) {
+        assert(context);
+
+        EDIT_FILES_FOREACH(i, context)
+                if (i->temp)
+                        (void) unlink(i->temp);
+}
+
+void edit_file_context_done(EditFileContext *context) {
+        assert(context);
+
+        EDIT_FILES_FOREACH(i, context) {
+                if (i->temp) {
+                        (void) unlink(i->temp);
+                        free(i->temp);
+                }
+
+                if (context->remove_parent) {
+                        _cleanup_free_ char *parent = NULL;
+                        int r;
+
+                        r = path_extract_directory(i->path, &parent);
+                        if (r < 0)
+                                log_error_errno(r, "Failed to extract directory from '%s': %m", i->path);
+
+                        /* No need to check if the dir is empty, rmdir does nothing if it is not the case. */
+                        (void) rmdir(parent);
+                }
 
-        EDIT_FILE_FOREACH(i, *ef) {
                 free(i->path);
-                free(i->tmp);
+                free(i->original_path);
+                strv_free(i->comment_paths);
         }
 
-        *ef = mfree(*ef);
+        context->files = mfree(context->files);
+        context->n_files = 0;
```

Comment:
Not a big matter, but we usually do not assign `NULL` or `0` in `xyz_done()`. But _done_ in `xyz_clear()`.